torch/_masked/__init__.py

```diff
@@ -988,17 +988,38 @@ def _output_mask(op, input: Tensor, *args, **kwargs) -> Tensor:
 
 
 def _combine_input_and_mask(op, input: Tensor, mask, *args) -> Tensor:
-    """Return input with masked-out elements eliminated for the given operations."""
-    if mask is None:
-        return input
-    canonical_mask = _input_mask(input, mask=mask)
-    if callable(op):
-        fill_value = _reduction_identity(op.__name__, input, *args)
-        return _where(canonical_mask, input, fill_value)
-    else:
-        raise ValueError(
-            f"_combine_input_and_mask expected masked operation (got {type(op).__name__} object)"
-        )
+    class Combine(torch.autograd.Function):
```

Comment:
Personal nit:
I think that the `.data()` and `_combine_input_and_mask()` as they are today are a bit confusing.
In particular, I feel like the `input: Tensor` that you get from `.data()` and pass in here is in an "invalid" state (the masked out values can contain arbitrary things that the user should really not look at.

I would suggest that we make this `_combine_input_and_mask` function take a mt as input directly and it will access the data directly.
I can see that there are cases where you don't care about the what the masked value contain, but these will be more rare and will be better expressed by a no-op identity for this function?

Does have to happen, just wanted to share.

Reply:
Thanks for the comment - definitely hear you on the confusion here and the undesirability of exposing `.get_data()` to other functions. As Basil pointed out, `_masked` functions are not necessarily meant to be only helpers for MaskedTensor (though we are the primary consumers of them at the moment. I think there are a few other files that use a function or two, but not necessarily the ones that we would be changing)

We would indeed need to change the functions like `sum`, `prod`, and so on to take in MTs as well. As a result, this is probably a bit out of scope for this PR and I'll table this as an improvement for the future if that s ounds g ood!